usr.bin/mail/mail.1
.\" Copyright (c) 1980, 1990 The Regents of the University of California.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)mail.1	6.19 (Berkeley) 7/27/91
.\"
.Dd July 27, 1991
.Dt MAIL 1
.Os BSD 4
.Sh NAME
.Nm mail
.Nd send and receive mail
.Sh SYNOPSIS
.Nm mail
.Op Fl iInv
.Op Fl s Ar subject
.Op Fl c Ar cc-addr
.Op Fl b Ar bcc-addr
.Ar to-addr...
.Nm mail
.Op Fl iInNv
.Fl f
.Op Ar name
.Nm mail
.Op Fl iInNv
.Op Fl u Ar user
.Sh INTRODUCTION
.Nm Mail
is a intelligent mail processing system, which has
a command syntax reminiscent of
.Xr \&ed 1
with lines replaced by messages.
.Pp
.Bl -tag -width flag
.It Fl v
Verbose mode.
The details of
delivery are displayed on the users terminal.
.It Fl i
Ignore tty interrupt signals.
This is
particularly useful when using
.Nm mail
on noisy phone lines.
.It Fl I
Forces mail to run in interactive mode even when
input isn't a terminal.
In particular, the
.Sq Ic \&~
special
character when sending mail is only active in interactive mode.
.It Fl n
Inhibits reading
.Pa /etc/mail.rc
upon startup.
.It Fl N
Inhibits the initial display of message headers
when reading mail or editing a mail folder.
.It Fl s
Specify subject on command line
(only the first argument after the
.Fl s
flag is used as a subject; be careful to quote subjects
containing spaces.)
.It Fl c
Send carbon copies to
.Ar list
of users.
.It Fl b
Send blind carbon copies to
.Ar list .
List should be a comma-separated list of names.
.It Fl f
Read in the contents of your
.Ar mbox
(or the specified file)
for processing; when you
.Ar quit  ,
.Nm mail
writes undeleted messages back to this file.
.It Fl u
Is equivalent to:
.Pp
.Dl mail -f /var/spool/mail/user
.El
.Ss Sending mail
To send a message to one or more people,
.Nm mail
can be invoked with arguments which are the names of people to
whom the mail will be sent.
You are then expected to type in
your message, followed
by an
.Sq Li control\-D
at the beginning of a line.
The section below
.Ar Replying to or originating mail ,
describes some features of
.Nm mail
available to help you compose your letter.
.Pp
.Ss Reading mail
In normal usage
.Nm mail
is given no arguments and checks your mail out of the
post office, then
prints out a one line header of each message found.
The current message is initially the first message (numbered 1)
and can be printed using the
.Ic print
command (which can be abbreviated
.Ql Ic p ) .
You can move among the messages much as you move between lines in
.Xr \&ed 1 ,
with the commands
.Ql Ic \&+
and
.Ql Ic \&\-
moving backwards and forwards, and
simple numbers.
.Pp
.Ss Disposing of mail.
After examining a message you can
.Ic delete
.Ql Ic d )
the message or
.Ic reply
.Ql Ic r )
to it.
Deletion causes the
.Nm mail
program to forget about the message.
This is not irreversible; the message can be
.Ic undeleted
.Ql Ic u )
by giving its number, or the
.Nm mail
session can be aborted by giving the
.Ic exit
.Ql Ic x )
command.
Deleted messages will, however, usually disappear never to be seen again.
.Pp
.Ss Specifying messages
Commands such as
.Ic print
and
.Ic delete
can be given a list of message numbers as arguments to apply
to a number of messages at once.
Thus
.Dq Li delete 1 2
deletes messages 1 and 2, while
.Dq Li delete 1\-5
deletes messages 1 through 5.
The special name
.Ql Li \&*
addresses all messages, and
.Ql Li \&$
addresses
the last message; thus the command
.Ic top
which prints the first few lines of a message could be used in
.Dq Li top \&*
to print the first few lines of all messages.
.Pp
.Ss Replying to or originating mail.
You can use the
.Ic reply
command to
set up a response to a message, sending it back to the
person who it was from.
Text you then type in, up to an end-of-file,
defines the contents of the message.
While you are composing a message,
.Nm mail
treats lines beginning with the character
.Ql Ic \&~
specially.
For instance, typing
.Ql Ic \&~m
(alone on a line) will place a copy
of the current message into the response right shifting it by a tabstop
(see
.Em indentprefix
variable, below).
Other escapes will set up subject fields, add and delete recipients
to the message and allow you to escape to an editor to revise the
message or to a shell to run some commands.
(These options
are given in the summary below.)
.Pp
.Ss Ending a mail processing session.
You can end a
.Nm mail
session with the
.Ic quit
.Ql Ic q )
command.
Messages which have been examined go to your
.Ar mbox
file unless they have been deleted in which case they are discarded.
Unexamined messages go back to the post office.
(See the
.Fl f
option above).
.Pp
.Ss Personal and systemwide distribution lists.
It is also possible to create a personal distribution lists so that,
for instance, you can send mail to
.Dq Li cohorts
and have it go
to a group of people.
Such lists can be defined by placing a line like
.Pp
.Dl alias cohorts bill ozalp jkf mark kridle@ucbcory
.Pp
in the file
.Pa \&.mailrc
in your home directory.
The current list of such aliases can be displayed with the
.Ic alias
command in
.Nm mail  .
System wide distribution lists can be created by editing
.Pa /etc/aliases ,
see
.Xr aliases  5
and
.Xr sendmail  8  ;
these are kept in a different syntax.
In mail you send, personal aliases will be expanded in mail sent
to others so that they will be able to
.Ic reply
to the recipients.
System wide
.Ic aliases
are not expanded when the mail is sent,
but any reply returned to the machine will have the system wide
alias expanded as all mail goes through
.Xr sendmail  .
.Pp
.Ss Network mail (ARPA, UUCP, Berknet)
See
.Xr mailaddr 7
for a description of network addresses.
.Pp
.Nm Mail
has a number of options which can be set in the
.Pa .mailrc
file to alter its behavior; thus
.Dq Li set askcc
enables the
.Ar askcc
feature.
(These options are summarized below.)
.Sh SUMMARY
(Adapted from the `Mail Reference Manual')
.Pp
Each command is typed on a line by itself, and may take arguments
following the command word.
The command need not be typed in its
entirety \- the first command which matches the typed prefix is used.
For commands which take message lists as arguments, if no message
list is given, then the next message forward which satisfies the
command's requirements is used.
If there are no messages forward of
the current message, the search proceeds backwards, and if there are no
good messages at all,
.Nm mail
types
.Dq Li No applicable messages
and
aborts the command.
.Bl -tag -width delete
.It Ic \&\-
Print out the preceding message.
If given a numeric
argument
.Ar n  ,
goes to the
.Ar n Ns 'th
previous message and prints it.
.It Ic \&?
Prints a brief summary of commands.
.It Ic \&!
Executes the shell
(see
.Xr sh 1
and
.Xr csh 1 )
command which follows.
.It Ic Print
.Pq Ic P
Like
.Ic print
but also prints out ignored header fields.
See also
.Ic print ,
.Ic ignore
and
.Ic retain .
.It Ic Reply
.Pq Ic R
Reply to originator.
Does not reply to other
recipients of the original message.
.It Ic Type
.Pq Ic T
Identical to the
.Ic Print
command.
.It Ic alias
.Pq Ic a
With no arguments, prints out all currently-defined aliases.
With one
argument, prints out that alias.
With more than one argument, creates
a new alias or changes an old one.
.It Ic alternates
.Pq Ic alt
The
.Ic alternates
command is useful if you have accounts on several machines.
It can be used to inform
.Nm mail
that the listed addresses are really you.
When you
.Ic reply
to messages,
.Nm mail
will not send a copy of the message to any of the addresses
listed on the
.Ic alternates
list.
If the
.Ic alternates
command is given with no argument, the current set of alternate
names is displayed.
.It Ic chdir
.Pq Ic c
Changes the user's working directory to that specified, if given.
If
no directory is given, then changes to the user's login directory.
.It Ic copy
.Pq Ic co
The
.Ic copy
command does the same thing that
.Ic save
does, except that it does not mark the messages it
is used on for deletion when you quit.
.It Ic delete
.Pq Ic d
Takes a list of messages as argument and marks them all as deleted.
Deleted messages will not be saved in
.Ar mbox  ,
nor will they be available for most other commands.
.It Ic dp
(also
.Ic dt )
Deletes the current message and prints the next message.
If there is no next message,
.Nm mail
says
.Dq Li "at EOF" .
.It Ic edit
.Pq Ic e
Takes a list of messages and points the text editor at each one in
turn.
On return from the editor, the message is read back in.
.It Ic exit
.Pf ( Ic ex
or
.Ic x )
Effects an immediate return to the Shell without
modifying the user's system mailbox, his
.Ar mbox
file, or his edit file in
.Fl f  .
.It Ic file
.Pq Ic fi
The same as
.Ic folder  .
.It Ic folders
List the names of the folders in your folder directory.
.It Ic folder
.Pq Ic fo
The
.Ic folder
command switches to a new mail file or folder.
With no
arguments, it tells you which file you are currently reading.
If you give it an argument, it will write out changes (such
as deletions) you have made in the current file and read in
the new file.
Some special conventions are recognized for
the name.
# means the previous file, % means your system
mailbox, %user means user's system mailbox, & means
your
.Ar mbox
file, and
\&+\&folder means a file in your folder
directory.
.It Ic from
.Pq Ic f
Takes a list of messages and prints their message headers.
.It Ic headers
.Pq Ic h
Lists the current range of headers, which is an 18\-message group.
If
a
.Ql \&+
argument is given, then the next 18\-message group is printed, and if
a
.Ql \&\-
argument is given, the previous 18\-message group is printed.
.It Ic help
A synonym for
.Ic \&?
.It Ic hold
.Pf ( Ic ho ,
also
.Ic preserve )
Takes a message list and marks each
message therein to be saved in the
user's system mailbox instead of in
.Ar mbox  .
Does not override the
.Ic delete
command.
.It Ic ignore
Add the list of header fields named to the
.Ar ignored list
Header fields in the ignore list are not printed
on your terminal when you print a message.
This
command is very handy for suppression of certain machine-generated
header fields.
The
.Ic Type
and
.Ic Print
commands can be used to print a message in its entirety, including
ignored fields.
If
.Ic ignore
is executed with no arguments, it lists the current set of
ignored fields.
.It Ic mail
.Pq Ic m
Takes as argument login names and distribution group names and sends
mail to those people.
.It Ic mbox
Indicate that a list of messages be sent to
.Ic mbox
in your home directory when you quit.
This is the default
action for messages if you do
.Em not
have the
.Ic hold
option set.
.It Ic next
.Pq Ic n
like
.Ic \&+
or
.Tn CR )
Goes to the next message in sequence and types it.
With an argument list, types the next matching message.
.It Ic preserve
.Pq Ic pre
A synonym for
.Ic hold  .
.It Ic print
.Pq Ic p
Takes a message list and types out each message on the user's terminal.
.It Ic quit
.Pq Ic q
Terminates the session, saving all undeleted, unsaved messages in
the user's
.Ar mbox
file in his login directory, preserving all messages marked with
.Ic hold
or
.Ic preserve
or never referenced
in his system mailbox, and removing all other messages from his system
mailbox.
If new mail has arrived during the session, the message
.Dq Li "You have new mail"
is given.
If given while editing a
mailbox file with the
.Fl f
flag, then the edit file is rewritten.
A return to the Shell is
effected, unless the rewrite of edit file fails, in which case the user
can escape with the
.Ic exit
command.
.It Ic reply
.Pq Ic r
Takes a message list and sends mail to the sender and all
recipients of the specified message.
The default message must not be deleted.
.It Ic respond
A synonym for
.Ic reply  .
.It Ic retain
Add the list of header fields named to the
.Ar retained list
Only the header fields in the retain list
are shown on your terminal when you print a message.
All other header fields are suppressed.
The
.Ic Type
and
.Ic Print
commands can be used to print a message in its entirety.
If
.Ic retain
is executed with no arguments, it lists the current set of
retained fields.
.It Ic save
.Pq Ic s
Takes a message list and a filename and appends each message in
turn to the end of the file.
The filename in quotes, followed by the line
count and character count is echoed on the user's terminal.
.It Ic set
.Pq Ic se
With no arguments, prints all variable values.
Otherwise, sets
option.
Arguments are of the form
.Ar option=value
(no space before or after =) or
.Ar option .
Quotation marks may be placed around any part of the assignment statement to
quote blanks or tabs, i.e.
.Dq Li "set indentprefix=\*q->\*q"
.It Ic saveignore
.Ic Saveignore
is to
.Ic save
what
.Ic ignore
is to
.Ic print
and
.Ic type  .
Header fields thus marked are filtered out when
saving a message by
.Ic save
or when automatically saving to
.Ar mbox  .
.It Ic saveretain
.Ic Saveretain
is to
.Ic save
what
.Ic retain
is to
.Ic print
and
.Ic type  .
Header fields thus marked are the only ones saved
with a message when saving by
.Ic save
or when automatically saving to
.Ar mbox  .
.Ic Saveretain
overrides
.Ic saveignore  .
.It Ic shell
.Pq Ic sh
Invokes an interactive version of the shell.
.It Ic size
Takes a message list and prints out the size in characters of each
message.
.It Ic source
The
.Ic source
command reads
commands from a file.
.It Ic top
Takes a message list and prints the top few lines of each.
The number of
lines printed is controlled by the variable
.Ic toplines
and defaults to five.
.It Ic type
.Pq Ic t
A synonym for
.Ic print  .
.It Ic unalias
Takes a list of names defined by
.Ic alias
commands and discards the remembered groups of users.
The group names
no longer have any significance.
.It Ic undelete
.Pq Ic u
Takes a message list and marks each message as
.Ic not
being deleted.
.It Ic unread
.Pq Ic U
Takes a message list and marks each message as
.Ic not
having been read.
.It Ic unset
Takes a list of option names and discards their remembered values;
the inverse of
.Ic set  .
.It Ic visual
.Pq Ic v
Takes a message list and invokes the display editor on each message.
.It Ic write
.Pq Ic w
Similar to
.Ic save  ,
except that
.Ic only
the message body
.Pq Ar without
the header) is saved.
Extremely useful for such tasks as sending and receiving source
program text over the message system.
.It Ic xit
.Pq Ic x
A synonym for
.Ic exit  .
.It Ic z
.Nm Mail
presents message headers in windowfuls as described under the
.Ic headers
command.
You can move
.Nm mail Ns 's
attention forward to the next window with the
.Ic \&z
command.
Also, you can move to the previous window by using
.Ic \&z\&\-  .
.El
.Ss Tilde/Escapes
.Pp
Here is a summary of the tilde escapes,
which are used when composing messages to perform
special functions.
Tilde escapes are only recognized at the beginning
of lines.
The name
.Dq Em tilde\ escape
is somewhat of a misnomer since the actual escape character can be set
by the option
.Ic escape .
.Bl -tag -width Ds
.It Ic \&~! Ns Ar command
Execute the indicated shell command, then return to the message.
.It Ic \&~b Ns Ar name ...
Add the given names to the list of carbon copy recipients but do not make
the names visible in the Cc: line ("blind" carbon copy).
.It Ic \&~c Ns Ar name ...
Add the given names to the list of carbon copy recipients.
.It Ic \&~d
Read the file
.Dq Pa dead.letter
from your home directory into the message.
.It Ic \&~e
Invoke the text editor on the message collected so far.
After the
editing session is finished, you may continue appending text to the
message.
.It Ic \&~f Ns Ar messages
Read the named messages into the message being sent.
If no messages are specified, read in the current message.
Message headers currently being ignored (by the
.Ic ignore
or
.Ic retain
command) are not included.
.It Ic \&~F Ns Ar messages
Identical to
.Ic \&~f ,
except all message headers are included.
.It Ic \&~h
Edit the message header fields by typing each one in turn and allowing
the user to append text to the end or modify the field by using the
current terminal erase and kill characters.
.It Ic \&~m Ns Ar messages
Read the named messages into the message being sent, indented by a
tab or by the value of
.Ar indentprefix  .
If no messages are specified,
read the current message.
Message headers currently being ignored (by the
.Ic ignore
or
.Ic retain
command) are not included.
.It Ic \&~M Ns Ar messages
Identical to
.Ic \&~m ,
except all message headers are included.
.It Ic \&~p
Print out the message collected so far, prefaced by the message header
fields.
.It Ic \&~q
Abort the message being sent, copying the message to
.Dq Pa dead.letter
in your home directory if
.Ic save
is set.
.It Ic \&~r Ns Ar filename
Read the named file into the message.
.It Ic \&~s Ns Ar string
Cause the named string to become the current subject field.
.It Ic \&~\&t Ns Ar name ...
Add the given names to the direct recipient list.
.It Ic \&~\&v
Invoke an alternate editor (defined by the
.Ev VISUAL
option) on the
message collected so far.
Usually, the alternate editor will be a
screen editor.
After you quit the editor, you may resume appending
text to the end of your message.
.It Ic \&~w Ns Ar filename
Write the message onto the named file.
.It Ic \&~\&| Ns Ar command
Pipe the message through the command as a filter.
If the command gives
no output or terminates abnormally, retain the original text of the
message.
The command
.Xr fmt 1
is often used as
.Ic command
to rejustify the message.
.It Ic \&~: Ns Ar mail-command
Execute the given mail command.
Not all commands, however, are allowed.
.It Ic \&~~ Ns Ar string
Insert the string of text in the message prefaced by a single ~.
If
you have changed the escape character, then you should double
that character in order to send it.
.El
.Ss Mail Options
Options are controlled via
.Ic set
and
.Ic unset
commands.
Options may be either binary, in which case it is only
significant to see whether they are set or not; or string, in which
case the actual value is of interest.
The binary options include the following:
.Bl -tag -width append
.It Ar append
Causes messages saved in
.Ar mbox
to be appended to the end rather than prepended.
This should always be set (perhaps in
.Pa /etc/mail.rc ) .
.It Ar ask
Causes
.Nm mail
to prompt you for the subject of each message you send.
If
you respond with simply a newline, no subject field will be sent.
.It Ar askcc
Causes you to be prompted for additional carbon copy recipients at the
end of each message.
Responding with a newline indicates your
satisfaction with the current list.
.It Ar autoprint
Causes the
.Ic delete
command to behave like
.Ic dp
\- thus, after deleting a message, the next one will be typed
automatically.
.It Ar debug
Setting the binary option
.Ar debug
is the same as specifying
.Fl d
on the command line and causes
.Nm mail
to output all sorts of information useful for debugging
.Nm mail  .
.It Ar dot
The binary option
.Ar dot
causes
.Nm mail
to interpret a period alone on a line as the terminator
of a message you are sending.
.It Ar hold
This option is used to hold messages in the system mailbox
by default.
.It Ar ignore
Causes interrupt signals from your terminal to be ignored and echoed as
@'s.
.It Ar ignoreeof
An option related to
.Ar dot
is
.Ar ignoreeof
which makes
.Nm mail
refuse to accept a control-d as the end of a message.
.Ar Ignoreeof
also applies to
.Nm mail
command mode.
.It Ar metoo
Usually, when a group is expanded that contains the sender, the sender
is removed from the expansion.
Setting this option causes the sender
to be included in the group.
.It Ar noheader
Setting the option
.Ar noheader
is the same as giving the
.Fl N
flag on the command line.
.It Ar nosave
Normally, when you abort a message with two
.Tn RUBOUT
(erase or delete)
.Nm mail
copies the partial letter to the file
.Dq Pa dead.letter
in your home directory.
Setting the binary option
.Ar nosave
prevents this.
.It Ar Replyall
Reverses the sense of
.Ic reply
and
.Ic Reply
commands.
.It Ar quiet
Suppresses the printing of the version when first invoked.
.It Ar verbose
Setting the option
.Ar verbose
is the same as using the
.Fl v
flag on the command line.
When mail runs in verbose mode,
the actual delivery of messages is displayed on he users
terminal.
.El
.Ss Option String Values
.Bl -tag -width Va
.It Ev EDITOR
Pathname of the text editor to use in the
.Ic edit
command and
.Ic \&~e
escape.
If not defined, then a default editor is used.
.It Ev LISTER
Pathname of the directory lister to use in the
.Ic folders
command.
Default is
.Pa /bin/ls .
.It Ev PAGER
Pathname of the program to use in the
.Ic more
command or when
.Ic crt
variable is set.
The default paginator
.Xr more 1
is used if this option is not defined.
.It Ev SHELL
Pathname of the shell to use in the
.Ic \&!
command and the
.Ic \&~!
escape.
A default shell is used if this option is
not defined.
.It Ev VISUAL
Pathname of the text editor to use in the
.Ic visual
command and
.Ic \&~v
escape.
.It Va crt
The valued option
.Va crt
is used as a threshold to determine how long a message must
be before
.Ev PAGER
is used to read it.
If
.Va crt
is set without a value,
then the height of the terminal screen stored in the system
is used to compute the threshold (see
.Xr stty 1 ) .
.It Ar escape
If defined, the first character of this option gives the character to
use in the place of ~ to denote escapes.

named/CHANGES
This is version 4.8.1 of bind.  It is a test release that includes
version 4.8 with fixes, asynchronous zone transfer and better reload
capabilities.  Although it is currently described as a test release,
it is believed to be reasonably stable and more usable than the currently-
released version, 4.8.  The changes of note are:

  o The asynchronous zone transfer code previously posted to the bind
    mailing list has been integrated, completed and tested.  There are
    a number of changes from the version posted, including fixes to
    allow top-level domains to work and a simplification of the timer
    code.

  o The code for reloading the server has been changed so that only
    primary zones master files that have changed are reloaded.  The
    cache and secondary zones are not flushed, and the sortlist, domain,
    etc. are reset to correspond to the boot file contents.

  o Several bugs have been fixed: the name "*" is not interpreted as
    a wildcard in cached zones, only in primary zones.  Secondary servers
    no longer decrement the time-to-live of records by the time since
    they verified the zone with the master; as a result, they never
    hand out nameserver referrals with too short a TTL to be usable.
    A bug was fixed that caused secondary servers with out-of-date
    zones to return empty answers between the actual expiration time
    and the next timeout.

There are several other bugs that have been reported but have not yet
been fixed.  In addition, the next regular release of named will
support negative caching, but this has not been integrated.

I would appreciate receiving feedback on this release; in particular,
problems (or lack of problems) when installing on various systems.
I attempted to update the SysV code when integrating the zone-transfer,
but haven't tested it.

The notes from version 4.8 follow.

----------
Welcome to version 4.8 of bind.

There have been several changes to the named boot file (/etc/named.boot)
of which you should be aware.  The "domain" line for each zone is no longer
needed, but one such line may still be used to specify a default domain
to be used for queries containing names with only a single component.
The term "suffixes", which was added in version 4.7alpha, has been removed.

The manual page on named (named.8) has been updated to reflect all
these changes.  Please read this and look at the example files
before installation.  You should also note the changes in the
resolver code to support non-fully-qualified addresses and per-user
host aliases.  See hostname(7) for an overview.  Two new routines
have been added to the resolver library since the last test release:
res_query formulates a query, sends it, waits for a response and does
preliminary error checking; res_search implements the search rules
of gethostbyname using res_query.

The MX lookup routine in sendmail has been modified to use res_search.  
Also, dn_skip takes an additional parameter and has been renamed
to dn_skipname.  While old sendmail binaries will work with the new
version of bind, because of these changes, it is desirable to install
new sendmail sources and recompile sendmail.  Do not rebuild sendmail 
from old sources.   The new sendmail is on ucbarpa.Berkeley.EDU for
anonymous FTP from pub/4.3/sendmail.MX.tar and pub/4.3/sendmail.MX.tar.Z.

There have been numerous changes to named, fixing most of the known
bugs that can be fixed without major structural changes in the server.
Several server configurations that failed before should now work.
Certain robustness problems have been fixed, in particular bounds-
checking when processing incoming packets.  Two changes have been made
in preparation for negative caching: SOA records are sent in the authority
section in negative responses with NXDOMAIN set, and a bug was fixed that
caused confusion and repeated requests if a response had no error, no answer
and an SOA in the authority section.  As such responses are already sent
by other servers, and will be sent by the next release of BIND, it is
important that all sites upgrade to this version as quickly as possible.

The root "hint" cache and cache file remain the largest problem area,
along with named's naivete in accepting bogus server's data.
These will be addressed in the next release, along with asynchronous
zone transfers, intelligent reloading of zone files, faster startup,
and caching of negative responses.

This version (4.8) will replace the last officially released version (4.5).
Version 4.5 has a serious bug that causes the generation of a continuous
stream of bogons to the root domain servers (bogus queries with the query
response bit set and possibly garbage for nsid and rcode).  It is imperative
that these versions of named be replaced as fast as possible.  We urge you to
field 4.8 quickly, for the sake of the root domain servers.

	Mike Karels
	Jean Wood
	bind@ucbarpa.Berkeley.EDU


misc/mail.c
#

/*
 * mail command usage
 *	mail [-yn]
 *		prints your mail
 *	mail people
 *		sends standard input to people
 *
 *	mail -r machine user people
 *		sends mail from the network
 */

#define	SIGINT	2
#define DIRECT	040000
#define RMAILCMD "/usr/net/bin/sendmail"

struct inode {
	char minor;
	char major;
	int inumber;
	int flags;
	char nlinks;
	char uid;
	char gid;
	char size0;
	int size1;
	int addr[8];
	int actime[2];
	int modtime[2];
} inode;

char	lettmp[] "/tmp/maXXXXX";
char	preptmp[] "/tmp/mbXXXXX";
int	pwfil;
int	chew;
int	errs;

main(argc, argv)
char **argv;
{
	register int me;
	extern int fout;
	int uf, delexit();
	char namebuf[20];

	mktemp(lettmp);
	mktemp(preptmp);
	unlink(lettmp);
	unlink(preptmp);
	me = getuid();
	if (getname(me, namebuf) < 0) {
		printf("Who are you?\n");
		delexit(1);
	}
	if (argc < 2)
		goto hitit;
	for (argc--, argv++; argc > 0 && argv[0][0] == '-'; argc--, argv++)
	switch(argv[0][1]) {
		register char *cp, *np;

		case 'y':
		case 'n':
			argc++, argv--;
hitit:
			printmail(argc, argv, namebuf);
			delexit(0);

		case 'r':
			if (argc < 2)
				continue;
			if (!equal("network", namebuf) && me != 0) {
				printf("Nice try!\n");
				delexit(1);
			}
			chew++;
			np = namebuf;
			for (cp = argv[1]; *cp; cp++)
				*np++ = *cp;
			argc--, argv++;
			*np++ = ':';
			for (cp = argv[1]; *cp; cp++)
				*np++ = *cp;
			*np++ = 0;
			argc--, argv++;
			continue;
	}
	if ((signal(SIGINT, 01) & 01) == 0)
		signal(SIGINT, delexit);
	unlink(lettmp);
	fout = creat(lettmp, 0600);
	if (fout < 0) {
		fout = 1;
		perror(lettmp);
		delexit(1);
	}
	argc++, argv--;
	bulkmail(argc, argv, namebuf);
	delexit(0);
}

printmail(argc, argv, name)
char **argv;
char *name;
{
	extern int fin, fout;
	register n, c, f;
	char *mname;

	mname = cat("/usr/mail/", name);
	if (stat(mname, &inode)>=0 && inode.nlinks==1 &&
		fopen(mname, &fin)>=0 && (c = getchar())) {
		putchar(c);
		getput();
		close(fin);
		c = 'y';
		if (argc<2) {
			if (ttyn(0)!='x') {
				printf("Save?");
				fin = 0;
				c = getchar();
			}
		} else
			c = argv[1][1];
		if (!any(c, "yn"))
			delexit(0);
		if (c == 'y') {
			if (accesss("mbox")) {
				printf("Saved mail in 'mbox'\n");
				prepend(mname, "mbox", getuid());
				unlink(mname);
			} else
				printf("In wrong directory\n");
		} else
			unlink(mname);
	} else
		printf("No mail.\n");
}

bulkmail(argc, argv, from)
char **argv, *from;
{
	extern int fin, fout;
	register int c;
	int tbuf[2], ttyn1;

	fin = 0;
	(&fin)[1] = 0;
	time(tbuf);
	ttyn1 = ttyn(1);
	if (ttyn1 < 033) {
		ttyn1 =+ 'a' - 1;
		ttyn1 =<< 8;
		ttyn1 =| '^';
	}
	printf("From %s  tty%c  %s", from, ttyn1, ctime(tbuf));
	if (chew)
		do {
			c = getchar();
		} while (c != '\n' && c != 0);
	getput();
	putchar('\n');
	flush();
	close(fout);
	while (--argc > 0)
		sendto(*++argv);
	delexit(errs);
}

sendto(person)
char *person;
{
	static int saved;
	extern int fout, fin;
	register char *filep;
	register int him;

	if ((person[0] == 'y' || person[0] == 'Y') && person[1] == ':')
		person += 2;
	if ((person[0] | ' ') == 'c' && (person[1] | ' ') == 'o' &&
	    (person[2] | ' ') == 'r' && (person[3] | ' ') == 'y' &&
	    person[4] == ':')
		person += 5;
	if (any(':', person) || person[0] == 'm' && person[1] == 's' &&
	    person[2] == 'g' && person[3] == 's' && person[4] == 0) {
		int i = fork();
		int s;

		if (i < 0) {
			perror("fork");
			goto assback;
		}
		if (i == 0) {
			close(0);
			open(lettmp, 0);
			if (any(':', person)) {
				execl(RMAILCMD,"sendmail", person, 0);
				execl("/usr/bin/sendmail", "sendmail", person, 0);
				execl("/bin/sendmail", "sendmail", person, 0);
				perror("sendmail");
			} else {
				execl("/usr/new/msgs", "msgs", "-s", 0);
				execl("/usr/bin/msgs", "msgs", "-s", 0);
			}
			exit(12);
		}
		for (;;) {
			register int j = wait(&s);
			if (j == -1)
				goto assback;
			if (j == i)
				break;
		}
		if ((s & 0377) != 0 || (s >> 8) == 12)
			goto assback;
		return;
	}

	if ((him = getuserid(person)) == -1) {
assback:
		fout = 1;
		flush();
		printf("Can't send to %s.\n", person);
		errs++;
		if (ttyn(0)!='x' && saved==0) {
			saved++;
			if (accesss("dead.letter")) {
				printf("Letter saved in 'dead.letter'\n");
				prepend(lettmp, "dead.letter", getuid());
			} else
				printf("In wrong directory\n");
		}
		return;
	}
	prepend(lettmp, filep=cat("/usr/mail/", person), him);
}

prepend(from, to, own)
char *from, *to;
{
	extern int fin, fout;
	register int sig;

	unlink(preptmp);
	if (fcreat(preptmp, &fout) < 0) {
		fout = 1;
		perror("mail");
		delexit(1);
	}
	chmod(preptmp, 0600);
	if (fopen(from, &fin) < 0) {
		close(fout);
		fout = 1;
		perror("mail");
		unlink(preptmp);
		return(0);
	}
	getput();
	close(fin);
	fopen(to, &fin);
	getput();
	close(fin);
	flush();
	close(fout);
	sig = signal(SIGINT, 01);
	unlink(to);
	if (fcreat(to, &fout) < 0) {
		unlink(preptmp);
		fout = 1;
		signal(SIGINT, sig);
		return(0);
	}
	chmod(to, 0600);
	chown(to, own);
	if(stat(to, &inode) < 0 || inode.nlinks != 1) {
			close(fout);
			fout = 1;
			unlink(preptmp);
			signal(SIGINT, sig);
			return(0);
	}
	if (fopen(preptmp, &fin) < 0) {
		fout = 1;
		perror("mail");
		signal(SIGINT, sig);
		errs++;
		return(0);
	}
	getput();
	flush();
	close(fout);
	close(fin);
	fout = 1;
	signal(SIGINT, sig);
	return(1);
}

delexit(ex)
{
	unlink(lettmp);
	unlink(preptmp);
	exit(ex);
}

equal(as1, as2)
{
	register char *s1, *s2;

	s1 = as1;
	s2 = as2;
	while (*s1++ == *s2)
		if (*s2++ == 0)
			return(1);
	return(0);
}

cat(ap1, ap2)
char *ap1, *ap2;
{
	register char *p1, *p2;
	static char fn[32];

	p1 = ap1;
	p2 = fn;
	while (*p2++ = *p1++);
	p2--;
	p1 = ap2;
	while (*p2++ = *p1++);
	return(fn);
}

getput()
{
	extern int errno;
	register c;

	while(c = getchar()) {
		errno = 0;
		putchar(c);
		if(errno) {
			perror("mail");
			delexit(1);
		}
	}
}

accesss(s1)
{
	if (access(".", 2) != -1 && (stat(s1, &inode)<0 || access(s1, 2)==0))
		return(1);
	return(0);
}

any(c, str)
	char *str;
{
	register char *f;

	f = str;
	while (*f)
		if (c == *f++)
			return(1);
	return(0);
}
